LSM: Implement the block cache in Grid

ifreund opened the pull request:
see individual commit messages for further detail.

jorangreef committed:
LSM: Create SetAssociativeCache (WIP, happy holidays)

The beginning of a set-associative cache for caching billions of values,
with an Nth-chance CLOCK replacement policy for very low miss rates.

This is a work-in-progress at a VERY rough early stage, more to come!

The layout is configurable to benchmark and evaluate different layouts.

The big idea is to separate the values (8-65536 bytes), the tags
(1-2 bytes per value), the CLOCK counters (1-2 bits per value),
and the CLOCK hands (2-4 bits per set of 2,4 or 16 values).

This means we can then eliminate the padding that would otherwise be
introduced by the 2-4 bit CLOCK hands per set, so that we can keep
everything cache-line aligned and also improve the odds that most
metadata (counters, hands) will reside in the CPU cache.

For example, a 64 GiB 16-way cache for 64 KiB blocks, with 8-bit tags
and 2-bit CLOCK counters would have these contiguous allocations:

1048576 x 64 KiB blocks                       =  64 GiB
-------------------------------------------------------
1048576 x 8-bit SIMD tags                     =   1 MiB
1048576 x 2-bit CLOCK counters                = 256 KiB
1048576 / 16 ways * 4-bit CLOCK hands per set =  32 KiB

The key insight is that most of the time, we're only working in the
metadata zone, comparing against tags and checking CLOCK counters
(using the CLOCK counter also as a valid bit), and only very occasionally
(if the SIMD tag happens to match) having to do a cache miss into the
large blocks zone to compare against the actual key stored in the value.

The reason we don't support an 8-way set-associative cache is because the
CLOCK hands per set would then no longer be cache-line aligned. 16-way is
also a sweet spot for 128-bit SSE SIMD, which is widely supported.

Under the hood, this uses a shiny new PackedUnsignedIntegerArray to pack
the small (less than 8-bit) CLOCK counters and hands into 64-bit words,
without padding, little-endian only.

Happy holidays!

ifreund committed:
LSM: Tweak SetAssociativeCache API, implement eviction

The new get_or_put() API makes it possible to avoid a copy on insertion
if the key already exists.

ifreund committed:
LSM: Allow overriding alignment of SetAssociativeCache values

ifreund committed:
Journal: Move IOPS data structure to new file

This will be used by the LSM Forest's Grid as well.

ifreund committed:
LSM: Implement the block cache in Grid

This requires the compaction iterators to do an additional copy in their
read callbacks but this is OK. The primary motivation for having the
read buffers owned by the Grid is to eliminate copies in the much more
common point lookup case.

ifreund committed:
LSM: Queue concurrent reads to the same address in Grid

ifreund committed:
LSM: Add safety counter to set associative cache eviction loop

jorangreef commented on src/lsm/set_associative_cache.zig:
```diff
@@ -0,0 +1,423 @@
+
+    //const before = 0b10110010;
+    //const result = get_bits(u2, before, 0);
+    //std.debug.print("\nbefore={b}\nresult={b}\n", .{ before, result});
```
```suggestion
```

jorangreef commented on src/lsm/set_associative_cache.zig:
```diff
@@ -0,0 +1,423 @@
+            errdefer allocator.free(clocks);
+
+            _ = Clocks;
+            _ = Counts;
```
Are these still necessary now that it's no longer WIP?

ifreund replied:
Nope, I just forgot to remove them.

jorangreef commented on src/lsm/grid.zig:
```diff
@@ -91,49 +138,108 @@ pub fn GridType(comptime Storage: type) type {
+            grid.reads.release(iop);
+
+            // Always iterate through the full list of pending reads here to ensure that all
+            // possible reads in the list are started, even in the precense of concurrent reads
```
```suggestion
            // possible reads in the list are started, even in the presence of concurrent reads
```

jorangreef commented on src/lsm/set_associative_cache.zig:
```diff
@@ -0,0 +1,423 @@
+
+        fn search(self: *Self, set: Set, key: Key) ?*align(value_alignment) Value {
+            for (set.tags) |tag, way| {
+                // TODO self.counts.get()
```
Is this TODO still necessary?

ifreund replied:
I didn't write this TODO, but I think it's about changing the `PackedUnsignedIntegerArray` type to store the slice so that we could store use `self.counts.get()` instead of `Counts.get(self.counts)` which seems like it would be nice to do sometime.

jorangreef commented on src/lsm/tree.zig:
```diff
@@ -1949,29 +1940,48 @@ pub fn TreeType(
+                        mem.copy(u8, it.blocks.next_tail().?, bytes_read);
+                    }
+
                     it.blocks.advance_tail();
```
Grouping the copy into the tail with the advance of the tail, to be together in the source would be better, as they must stay in lock step. It would also be good to deduplicate the memcopy into the tail, by handling the "copy remaining values" special case separately upfront, and thereafter asserting that there is a `next_tail`.

This change would then make the "copy remaining values" branch distinct from the tail copy, so that it does only what it needs to do to "make space", dealing only in terms of the head, not the tail as well.

jorangreef commented on src/lsm/tree.zig:
```diff
@@ -1949,29 +1940,48 @@ pub fn TreeType(
                 }
 
-                fn on_read(read: *Grid.Read) void {
+                fn on_read(read: *Grid.Read, bytes_read: BlockPtrConst) void {
```
```suggestion
                fn on_read(read: *Grid.Read, block: BlockPtrConst) void {
```

`bytes_read` evokes the `usize` `read()` syscall return value, and also here the block ptr is to the full block. We could say `block_read` or just `block` which is what it is.

To make this possible, `it.block` should also be renamed to be clearer that it's an index.

jorangreef commented on src/lsm/set_associative_cache.zig:
```diff
@@ -0,0 +1,423 @@
+
+            const clock_index = @divExact(set.offset, layout.ways);
+
+            const max_iterations = layout.ways * math.maxInt(Count);
```
```suggestion
            const clock_iterations_max = layout.ways * math.maxInt(Count);
```

So that the qualifier, in this case `max`, comes last.

jorangreef commented on src/lsm/set_associative_cache.zig:
```diff
@@ -0,0 +1,423 @@
+
+                Counts.set(self.counts, set.offset + way, count - 1);
+            }
+            assert(safety_counter < max_iterations);
```
We typically use `_max` values elsewhere as if they are inclusive, not exclusive. i.e. One would expect the condition here to be in terms of `<= iterations_max`.

ifreund replied:
I don't know of anywhere else we start loop variables from `1` instead of `0` though, which is the only reasonable way I see to make this check use `<=` instead of `<`. I don't see a problem with using `<` here, starting loop variables/counters from `0` feels most natural and when doing so a check with `<` is 100% expected.

jorangreef replied:
> I don't know of anywhere else we start loop variables from 1 instead of 0 though

We don't typically start loop variables from 1 because we use 0-based indexing with 1-based counts.

However, both `safety_count` and `clock_iterations_max` are not *indexes* but *counts*, i.e. both 1-based, hence `<=` is less surprising. Otherwise, `safety_count` should be renamed to `safety_index`.

ifreund replied:
That's a good point, changed it to start at 1 and use `<=`!

jorangreef commented on src/lsm/set_associative_cache.zig:
```diff
@@ -0,0 +1,423 @@
+
+            var way = Clocks.get(self.clocks, clock_index);
+            while (safety_counter < max_iterations + 1) : ({
+                way +%= 1;
```
It would be good to have a comptime assertion that the type returned by `Clocks.get` will wrap exactly after all W ways have been traversed.

jorangreef commented on src/lsm/set_associative_cache.zig:
```diff
@@ -0,0 +1,423 @@
+            const clock_index = @divExact(set.offset, layout.ways);
+
+            const max_iterations = layout.ways * math.maxInt(Count);
+            var safety_counter: usize = 0;
```
```suggestion
            var safety_count: usize = 0;
```

We use `safety_counter` elsewhere but I've been thinking we should consolidate on just `_count` as we use for other things throughout the LSM namespace.

jorangreef commented on src/lsm/set_associative_cache.zig:
```diff
@@ -0,0 +1,423 @@
+            found_existing: bool,
+        };
+
+        pub fn get_or_put(self: *Self, key: Key) GetOrPutResult {
```
For the object cache, we also need an unconditional put method that doesn't carry the additional overhead of a search and that doesn't need to make any function calls or assertions related to locking.

That would be for cases where we just want to update the latest value.

jorangreef commented on src/lsm/set_associative_cache.zig:
```diff
@@ -0,0 +1,423 @@
+                    .value_ptr = existing,
+                    .found_existing = true,
+                };
+            }
```
When acquiring a block buffer from the cache, it feels like this is work that would have already been done upfront earlier at the call site?

i.e. `start_read()` only seems to need a `value_ptr` slot that hasn't been locked?

ifreund replied:
Indeed, if we always remove the corresponding block from the cache when `free_set.release()` we don't need this search call here. We could keep it as an assertion if `verify = true` however.

I'm inclined to scrap the `get_or_put()` API for a `put_no_clobber()` API that assumes the inserted key does not already exist in the cache and asserts that fact if `lsm.verify = true`. I believe that would satisfy the requirements for the object cache as well. 

`get_or_put()` seemed like a nice idea as it's pretty much the most flexible API possible, but it seems we don't need that flexibility here :)

jorangreef commented on src/lsm/set_associative_cache.zig:
```diff
@@ -0,0 +1,423 @@
+        /// If the key is not in the cache, add it to the cache, evicting older entries if needed.
+        /// Never evicts keys for which locked() returns true.
+        /// The caller must guarantee that locked() returns true for less than layout.ways keys.
+        pub fn get_or_put_preserve_locked(
```
It would be great to have a unit test for this function, checking how the clock hand is advanced across calls to the function, and how locked slots are skipped.

jorangreef reviewed (commented):
Great to see this coming together.

jorangreef commented on src/lsm/set_associative_cache.zig:
```diff
@@ -0,0 +1,423 @@
+        fn search(self: *Self, set: Set, key: Key) ?*align(value_alignment) Value {
+            for (set.tags) |tag, way| {
+                // TODO self.counts.get()
+                const count = Counts.get(self.counts, set.offset + way);
```
I missed this yesterday, and for the first draft last year didn't think about it beyond branching on the tag first—but the `const count` assignment must be moved to be within the `if (tag == set.tag)` branch, with a separate load of the count to evaluate the `count > 0` branch only if necessary.

The reason is that the tag contains enough entropy to avoid the second branch most of the time, so we don't even need to lookup the relevant count.

ifreund replied:
There's nothing stopping llvm from reordering the `Counts.get()` call after the `tag == set.tag` comparison with the current code, and my intuition says it would do so. That said, I'm happy to change the code and make this explicit, good thought!

ifreund replied:
bbda2adfba88543a8af728dc05a93806f9af6464

ifreund committed:
LSM: Make desired control flow in SetAssociativeCache.search() explicit

ifreund committed:
LSM: remove SAC.get_or_put(), add SAC.put_no_clobber()

It turns out that in practice we can ensure that block addresses we want
to add to the cache do not already exist in the cache by removing the
addresses from the cache when we release them in the free set. This
avoids a call to search(), though we still do this for the assertion
if online verification is enabled.

ifreund committed:
LSM: Fix size of clocks array in SetAssociativeCache

ifreund committed:
LSM: Store words slice in packed int array struct

This makes the API more pleasant to use as one can do foo.get(i)
instead of Foo.get(&foo, i).

ifreund committed:
LSM: Add more SetAssociativeCache tests, fix bug

A typo in the PackedUnsignedIntegerArray implementation caused set()
to zero out neighboring elements in some cases. Further bugs are now
guarded against with a fuzz test.

ifreund committed:
LSM: Fix bug in SetAssociativeCache eviction

Previously if a slot had count 1 which was decremented to 0 we would not
immediately break the eviction loop.

ifreund committed:
LSM: Add SetAssociativeCache.remove()

This will allow the Grid to free up space in the cache when block
addresses are released.

ifreund committed:
LSM: Implement Grid.acquire() and Grid.release()

ifreund committed:
LSM: Limit write concurrency in Grid

ifreund committed:
LSM: Tighten assertions in Grid.write_block()

ifreund committed:
LSM: Use SIMD in SetAssociativeCache.search()

ifreund committed:
LSM: Fuzz test SetAssociativeCache SIMD code

jorangreef commented on src/lsm/grid.zig:
```diff
@@ -19,54 +21,140 @@ pub fn GridType(comptime Storage: type) type {
     const SuperBlock = SuperBlockType(Storage);
 
+    const cache_interface = struct {
+        inline fn address_from_block(block: [block_size]u8) u64 {
```
It would be safer to take a `BlockPtrConst` to:

1. make it explicit that we don't want a copy,
2. eliminate any optimization dependency on the compiler (or performance risk due to regression), and
3. provide more type safety if the caller accidentally copies the block before passing it in.

ifreund replied:
Yes I agree, though making this change would sprawl outside of `Grid.zig` to every `key_from_value()` function in our LSM implementation. We accept
```zig
comptime key_from_value: fn (Value) callconv(.Inline) Key,
```
as a parameter of `TreeType()` in `lsm/tree.zig`.

Your call whether we should make this change as part of this PR or as a follow up.

jorangreef replied:
"sprawl" is a great word.

Let's do it, as a follow up.

jorangreef commented on src/lsm/grid.zig:
```diff
@@ -19,54 +21,140 @@ pub fn GridType(comptime Storage: type) type {
+        inline fn address_from_block(block: [block_size]u8) u64 {
+            const header_bytes = block[0..@sizeOf(vsr.Header)];
+            const header = mem.bytesAsValue(vsr.Header, header_bytes);
+            return header.op;
```
```suggestion
            const address = header.op;
            assert(address > 0);
            return address;
```

jorangreef commented on src/lsm/grid.zig:
```diff
@@ -19,54 +21,140 @@ pub fn GridType(comptime Storage: type) type {
+        }
+
+        inline fn hash_address(address: u64) u64 {
+            return std.hash.Wyhash.hash(0, mem.asBytes(&address));
```
```suggestion
            assert(address > 0);
            return std.hash.Wyhash.hash(0, mem.asBytes(&address));
```

jorangreef commented on src/lsm/grid.zig:
```diff
@@ -19,54 +21,140 @@ pub fn GridType(comptime Storage: type) type {
+                const callback = write.callback;
+                write.* = undefined;
+                callback(write);
+            }
```
Nice wrapper!

jorangreef commented on src/lsm/grid.zig:
```diff
@@ -19,54 +21,140 @@ pub fn GridType(comptime Storage: type) type {
+        }
+
+        pub fn acquire(grid: *Grid) u64 {
+            // We will reject incoming data before it reaches this point
```
```suggestion
            // We will reject incoming data before it reaches the point
```

jorangreef commented on src/lsm/grid.zig:
```diff
@@ -78,11 +166,72 @@ pub fn GridType(comptime Storage: type) type {
+        fn write_block_callback(completion: *Storage.Write) void {
+            const iop = @fieldParentPtr(WriteIOP, "completion", completion);
+
+            iop.write.finish();
```
Could we have some interesting concurrency interactions here around the order of how we release? For example, where the caller, synchronously within the callback, immediately resubmits a write for the same address or using the same block? From the caller's point of view that should be valid but I think it would trip our address/block assertions.

At the same time, we don't want the callback to be able to issue I/O that pre-empts the pending queue, so the order we have is good.

I wonder if we could deal with this more explicitly—by setting a boolean flag before we call `iop.write.finish()` and clearing it immediately afterwards, and then within `write_block()`, if this flag is set we immediately push to the pending queue?

ifreund replied:
Nice catch, I think I've come up with a clean solution here: fdbdbecb4643d424ce6ded1764176f9109565bf0

jorangreef commented on src/lsm/grid.zig:
```diff
@@ -78,11 +166,72 @@ pub fn GridType(comptime Storage: type) type {
+                .block = block,
+            };
+
+            grid.start_write(write);
```
It would be great over here to assert that if the I/O is acquired and started immediately, that the write queue must then be empty, to be sure that no pending writes were pre-empted. This would help catch interesting concurrency bugs that might be triggered where the callback synchronously issues more I/O, or where the pending queue is not being processed as expected.

jorangreef commented on src/lsm/grid.zig:
```diff
@@ -78,11 +166,72 @@ pub fn GridType(comptime Storage: type) type {
+            {
+                var it = grid.read_iops.iterate();
+                while (it.next()) |iop| {
+                    assert(address != iop.reads.peek().?.address);
```
This means that a write also can't start until any reads for the same `address` complete or it would trip this assertion. However, this may be fine, since we won't overlap prefetching or reads at the same time as compaction (which is the only place we write to blocks) in the commit log ratchet mechanism.

jorangreef replied:
It would be good to do these asserts rather at the moment that we start the I/O, so that there's no gap from when we do the assertions and when we issue the I/O. This will then also match `start_read()`.

ifreund replied:
> This means that a write also can't start until any reads for the same `address` complete or it would trip this assertion. However, this may be fine, since we won't overlap prefetching or reads at the same time as compaction (which is the only place we write to blocks) in the commit log ratchet mechanism.

Yes, I realize that there may be some use-cases for this API where racing writes and reads aren't problematic, but as you say our use-case doesn't need to read and write the same address at the same time. If we find a use-case where we do need to support that it will be easy to remove this check.

> It would be good to do these asserts rather at the moment that we start the I/O, so that there's no gap from when we do the assertions and when we issue the I/O. This will then also match `start_read()`.

~~I can see the argument for that, but I think it would be even stronger to have the assertions at both places.~~

Whoops, that didn't make sense, moving these to `start_write()` makes sense. :+1:

jorangreef replied:
😄 I was actually going to suggest checking in both places. What was your thinking on why it didn't make sense?

ifreund replied:
I think I said that because I saw putting in both places as superfluous and came around to seeing `start_write()` as the obvious/ideal place to put it.

I don't see any reason to not have both though, so I've gone ahead and done that. I also added some missing parallel assertions on the read path ee5db252b5ccf245d5cae742fb19850f4d076a83

jorangreef commented on src/lsm/grid.zig:
```diff
@@ -91,49 +240,107 @@ pub fn GridType(comptime Storage: type) type {
+                while (iop.reads.pop()) |read| {
+                    assert(read.address == address);
+                    assert(read.checksum == checksum);
+                    read.finish(iop.block);
```
If the callback issues another read for the same `address`/`checksum` is there anything here that could go wrong?

ifreund replied:
There is in fact a pathological case that could cause this while loop to never terminate. Not something I would expect actually useful code to ever hit in practice, but I'll use the same pattern we now use for the write path. It's also nice to keep them more consistent.

ifreund replied:
fixed in 7fafc438c3d5cfd3225fe72f0f6b4171f253c42e

jorangreef replied:
Great, I thought _let me just ask the question_ and you'll find what I hadn't spotted.

jorangreef commented on src/lsm/grid.zig:
```diff
@@ -91,49 +240,107 @@ pub fn GridType(comptime Storage: type) type {
+                    read.finish(iop.block);
+                }
             } else {
-                read.grid.read_recovery_queue.push(read);
```
It would be good to error log which of the 3 reasons (invalid checksum or body or checksum mismatch) meant that we had to push to the recovery queue.

jorangreef replied:
When doing this, we should also first check if the checksum is valid, before we use the checksum to compare it to the read's checksum.

jorangreef commented on src/lsm/grid.zig:
```diff
@@ -91,49 +240,107 @@ pub fn GridType(comptime Storage: type) type {
+                return;
+            };
+
+            const block = grid.cache.put_no_clobber_preserve_locked(
```
Where do we check if the cache has the block we want to read, i.e. so we can skip the I/O to disk?

jorangreef replied:
This should be handled by `Grid.read_block()` so that the caller only has a single code path (read-through, write-through cache), without having to "lookaside" to the cache.

`Grid.write_block()` should also update the cache once the write is done, and then we just need to be careful with concurrency—how this interacts with inflight or queued reads.

ifreund replied:
As of 7fafc438c3d5cfd3225fe72f0f6b4171f253c42e the block cache will be checked before disk I/O is started.

I haven't yet looked into having `Grid.write_block()` write to the cache as well.

ifreund replied:
> `Grid.write_block()` should also update the cache once the write is done, and then we just need to be careful with concurrency—how this interacts with inflight or queued reads.

This is almost fine as we disallow reading and writing the same address concurrently. The only tricky thing I found has to do with the locked slots in the block cache. We need to reduce the maximum number of concurrent reads by one to ensure writes are always able to add the written block to the cache on completion.

Implemented in ad5f22d384e43ce6f305810a6fe4e9cb205ff86b

jorangreef replied:
Nice catch. Off-by-one errors 😄

jorangreef commented on src/lsm/grid.zig:
```diff
@@ -91,49 +240,107 @@ pub fn GridType(comptime Storage: type) type {
+            var copy = grid.reads_pending;
+            grid.reads_pending = .{};
+            while (copy.pop()) |read| {
+                grid.start_read(read);
```
What happens if these hit the cache and not disk once started?

ifreund replied:
With 7fafc438c3d5cfd3225fe72f0f6b4171f253c42e there is no chance of unbounded recursion or an infinite loop.

This required adding a `Grid.tick()` function though as I could not find a way to implement this robustly otherwise.

jorangreef reviewed (commented):
Great work @ifreund — very clean. I've gone through `Grid` and will tackle `Tree` and `SetAssociativeCache` next. Wanted to get this to you so long.

jorangreef commented on src/lsm/grid.zig:
```diff
@@ -19,54 +21,140 @@ pub fn GridType(comptime Storage: type) type {
+        write_iops: IOPS(WriteIOP, write_iops_max) = .{},
+
+        reads_pending: FIFO(Read) = .{},
+        read_iops: IOPS(ReadIOP, read_iops_max) = .{},
```
```suggestion
        write_iops: IOPS(WriteIOP, write_iops_max) = .{},
        write_queue: FIFO(Write) = .{},

        read_iops: IOPS(ReadIOP, read_iops_max) = .{},
        read_queue: FIFO(Read) = .{},
```

To line up, and also to match the order that we use them: first try to acquire from `write_iops`, otherwise push to the queue.

Using a noun can also be very useful. For example, as an adjective, `pending` is good for something like a boolean state. But for a queue, a noun is better because that can more easily be referred to directly in external docs/communication, e.g. "Grid's write queue is full" without having to be rephrased e.g. "Grid has no pending writes".

ifreund committed:
LSM: Fix code style issues in Grid

ifreund committed:
LSM: Handle Grid.write_block() call within write callback

ifreund committed:
LSM: Tighten write path assertions in Grid

ifreund committed:
LSM: Improve logging on corrupt reads in Grid

ifreund committed:
FIFO: add empty()

This reads more nicely than `foo.peek() == null`.

ifreund committed:
LSM: Check cache for block before reading from disk

This requires care to avoid unbounded recursion or potential infinite
loops.

The new recursion guard fixes an existing potential infinite loop in
read_block_callback() as well.

ifreund committed:
LSM: Tighten assertions in Grid.release() and Grid.write_block()

ifreund committed:
LSM: Add completed writes to the block cache in Grid

ifreund committed:
LSM: Fix and tighten assertions in Grid write path

ifreund committed:
LSM: Strengthen asserts forbidding concurrent reads/writes in Grid

jorangreef commented on src/lsm/set_associative_cache.zig:
```diff
@@ -0,0 +1,719 @@
+    }
+
+    if (layout.value_alignment) |alignment| {
+        assert(alignment > @alignOf(Value));
```
```suggestion
        assert(alignment >= @alignOf(Value));
```

It should be okay to explicitly provide `value_alignment` as the expected alignment of the `Value`.

jorangreef committed:
SAC: Rename value_size_max to values_size_max

jorangreef committed:
SAC: matches_bitmask() MatchesBitmask -> search_tags() Ways

Since a bitmask is the bitwise constant applied to a byte or word of bits,
whereas the result here is after the vector mask has been applied.

ifreund committed:
LSM: Remove unnecessary `pub` usage in Grid

jorangreef committed:
SAC: Polish

- Inline BitIterator.next().
- Make types in search_tags() more explicit.
- Declare variables closer to where they are used in put...().
- Set tags/counts/clocks in the same order as their fields on the SAC.
- Fix line lengths.

jorangreef commented on src/lsm/set_associative_cache.zig:
```diff
@@ -0,0 +1,726 @@
+
+            const clock_iterations_max = layout.ways * math.maxInt(Count);
+            var safety_count: usize = 1;
+            while (safety_count <= clock_iterations_max + 1) : ({
```
```suggestion
            var safety_count: usize = 0;
            const clock_iterations_max = layout.ways * math.maxInt(Count);
            while (safety_count <= clock_iterations_max) : ({
```

Subtracting 1 from both `safety_count` and `clock_iterations_max` should have the same effect?

jorangreef replied:
We could then do `else unreachable` on the loop as a simpler way to check that we always break from the loop, and so that we don't need the plus one to cater for the last iteration, which is a tricky case.

ifreund replied:
> Subtracting 1 from both `safety_count` and `clock_iterations_max` should have the same effect?

That is in fact what I originally had and I'd be happy to go back to that. I changed this because you preferred `<=` to `<` in the assert: https://github.com/coilhq/tigerbeetle/pull/85#discussion_r836293072

jorangreef replied:
What I'm proposing here is slightly different. Both counts remain counts (instead of index/count as before), and with the `else unreachable` branch there's no longer a need for the `+ 1` to handle the last iteration. i.e. Before there was an assert on the safety counter after the loop. We can drop that assertion entirely with this change.

ifreund replied:
Ah I see, that is indeed cleaner. Thanks!

jorangreef replied:
Anytime!

jorangreef commented on src/lsm/set_associative_cache.zig:
```diff
@@ -0,0 +1,726 @@
+                self.counts.set(set.offset + way, count);
+                if (count == 0) break; // Way has become free.
+            }
+            assert(safety_count <= clock_iterations_max);
```
```suggestion
            assert(safety_count > 0);
            assert(safety_count <= clock_iterations_max);
```

Since we expect at least one iteration, and want to be sure that we did at least enter the loop to find an empty way.

ifreund replied:
Why do we expect at least one iteration? I don't think we do...

jorangreef replied:
Because if we never enter the loop, we never check if the way is actually free.

jorangreef replied:
We could make this even harder to get wrong by asserting that the `count` for the way is `0` before setting the count.

ifreund replied:
We always expect to enter the loop, but we may not complete that first iteration and increment the safety count if e.g. the set is empty.

> We could make this even harder to get wrong by asserting that the `count` for the way is `0` before setting the count.

Good idea :+1:

jorangreef replied:
Ah, of course. `assert(safety_count > 0)` would not have worked. But testing for the effect of the break, i.e. asserting that the count is 0 is also better and shows what we're really testing for.

ifreund replied:
Pushed a commit implementing this and the other suggestion :)

jorangreef commented on src/lsm/set_associative_cache.zig:
```diff
@@ -0,0 +1,726 @@
+    );
+    if (log) SAC.inspect();
+
+    const Tag = meta.Int(.unsigned, layout.tag_bits);
```
We should test `search_tags()` exhaustively for different sized tags, the ones we intend supporting, as well as the different number of ways that we intend to support.

Otherwise, we're only testing the default layout.

jorangreef reviewed (commented):
This is really clean!

ifreund committed:
LSM: Allow explicit SAC value alignment equal to natural alignment

ifreund committed:
LSM: Test SAC.search_tags() for all supported tag_bits/ways layouts

ifreund committed:
LSM: Clean up/tighten assertions in SAC eviction

jorangreef committed:
SAC: Fix line lengths, thin out test code

- Fix line lengths
- Assert match count in reference.search_tags()
- Use aliases for expect(), expectEqual() to thin out test code
- Group test context with test case

jorangreef commented on src/lsm/set_associative_cache.zig:
```diff
@@ -0,0 +1,760 @@
+            var it = BitIterator(Ways){ .bits = ways };
+            while (it.next()) |way| {
+                const count = self.counts.get(set.offset + way);
+                if (count > 0 and equal(key_from_value(set.values[way]), key)) {
```
It would be good to improve the test coverage for the `equal()` condition here.

I was bashing the code and found that `if (count > 0 and (true or equal(key_from_value(set.values[way]), key))) {` still passed all the tests.

ifreund replied:
I added a second run of the same tests with a "hash" function that always returns 0 in 67bd687137864e1ca830e098c5eec1a91b8e49da. The tests now fail if that change is made.

jorangreef commented on src/lsm/set_associative_cache.zig:
```diff
@@ -0,0 +1,760 @@
+
+        pub fn reset(self: *Self) void {
+            mem.set(Tag, self.tags, 0);
+            mem.set(u64, self.counts.words, 0);
```
It would be great for the test to verify all the `tag/count/clock` memory explicitly.

For example, the count is critical and needs to be reset correctly, and some invalid counts do fail the tests. However, `mem.set(u64, self.counts.words, 1);` passed the tests and there are a few other counts that also pass.

ifreund replied:
Done!

jorangreef commented on src/lsm/set_associative_cache.zig:
```diff
@@ -0,0 +1,760 @@
+            comptime assert(math.maxInt(@TypeOf(way)) == layout.ways - 1);
+            comptime assert(@as(@TypeOf(way), math.maxInt(@TypeOf(way))) +% 1 == 0);
+
+            const clock_iterations_max = layout.ways * math.maxInt(Count);
```
The tests don't cover the whole valid range of `clock_iterations_max`. For example, if we reduce `clock_iterations_max` by 1 or 2 then the tests pass.

ifreund replied:
Turns out that the maximum number of iterations is actually `layout.ways * (math.maxInt(Count) - 1)`.

This is now covered by the tests.

jorangreef reviewed (approved):
Looks great and ready to merge. Please would you tune some of the test coverage to cover a few important cases that are presently not covered.

ifreund committed:
LSM: Test SetAssociativeCache hash collision handling

ifreund committed:
LSM: Test SetAssociativeCache evicition loop max iterations

We now hit upper bound on iterations in the evicition loop in the tests,
which revealed that this upper bound is actually lower than initially
thought.

ifreund committed:
LSM: Explicitly test initial SetAssociativeCache metadata

jorangreef commented:
Thanks @ifreund nice work.
